<div style="display: flex; align-items: center; gap: 20px; font-family: Arial, sans-serif;">
    <div style="margin-right: 30px;">
        <img src="../Sources/Logo.png" width="300" style="border-radius: 10px;"/>
    </div>
    <div style="line-height: 1.5;">
        <strong style="font-size: 18px; display: block; margin-bottom: 5px;">
            <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">
                Репетитор по информатике | Мамаев Алексей
            </a>
        </strong>
        <span style="font-size: 16px; color: #333;">
            Разбор <a href='https://euler.jakumo.org/problem/14-samaia-dlinnaia-posledovatelnost-kollatca' style="text-decoration: none; color: #1a73e8;">
            14-й задачи</a> проекта Эйлера
        </span>
        <span style="font-size: 16px; color: #333;">
            <br><br><br><br><br><br>ТГК: <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">@M4m4ev_Alex</a><br>
        </span>
    </div>
</div>

## Задача 14

Для натурального числа `n` определяется следующая последовательность:

$$
n \rightarrow \frac{n}{2},
\qquad \text{если } n \text{ чётное}
$$

$$
n \rightarrow 3n + 1,
\qquad \text{если } n \text{ нечётное}
$$

Например, если начать с числа `13`, получится последовательность:

```text
13 → 40 → 20 → 10 → 5 → 16 → 8 → 4 → 2 → 1
```

Она содержит `10` элементов.

Согласно гипотезе Коллатца, любая такая последовательность в конечном итоге достигает числа `1`, однако это утверждение до сих пор не доказано.

Необходимо определить, какое начальное число, меньшее `1 000 000`, порождает самую длинную последовательность Коллатца.

> Следующие элементы последовательности могут превышать `1 000 000`.

Ответ: 837799

## 1) Рекурсивное построение последовательности

В данном решении длина последовательности Коллатца вычисляется с помощью рекурсивной функции `func_1()`.

Функция последовательно применяет правило Коллатца к текущему числу и вызывает саму себя для следующего элемента последовательности. Каждый рекурсивный вызов добавляет единицу к общей длине.

После этого для всех начальных чисел, меньших заданной границы, вычисляется длина последовательности и сохраняется число, породившее самый длинный вариант.

### Алгоритм

1. Если текущее число равно `1`, функция возвращает длину `1`.
2. Если число чётное, выполняется переход к `n // 2`.
3. Если число нечётное, выполняется переход к `3n + 1`.
4. Каждый рекурсивный вызов увеличивает длину последовательности на единицу.
5. В функции `get_result_1()` перебираются все числа от `1` до `n - 1`.
6. Для каждого числа вычисляется длина его последовательности.
7. Если найденная длина больше текущего максимума, сохраняется новое начальное число.
8. После завершения перебора возвращается число, породившее самую длинную последовательность.

### Особенности

- Используется рекурсивная реализация правила Коллатца.
- Для каждого начального числа последовательность вычисляется заново.
- Код отличается простотой и напрямую повторяет условие задачи.
- Пространственная сложность одной проверки зависит от глубины рекурсии.

### Вывод

Рекурсивный подход позволяет наглядно описать построение последовательности Коллатца. Однако одинаковые участки последовательностей вычисляются многократно, поэтому при переборе большого количества начальных значений алгоритм выполняет много повторных операций.

In [22]:
from time import perf_counter # измерение времени выполнения алгоритма

In [23]:
def func_1(n):
    if n == 1:
        return 1
    if n % 2 == 0:
        return 1 + func_1(n // 2)
    return 1 + func_1(3 * n + 1)

def get_result_1(n: int) -> int:
    mx = 0
    cmx = 0
    for i in range(1, n):
        c = func_1(i)
        if c > cmx:
            cmx = c
            mx = i
    return mx


start = perf_counter()
print(get_result_1(1_000_000))
end = perf_counter()

print(f'{end - start:.5f}')

837799
19.64032


## 2) Рекурсия с кэшированием

Во втором решении используется тот же рекурсивный подход, однако функция `func_2()` дополнительно использует декоратор `@cache` из модуля `functools`.

Кэширование сохраняет длину уже вычисленных последовательностей. Если в дальнейшем алгоритм снова достигает того же числа, результат берётся из памяти без повторного вычисления.

Для поиска начального числа с максимальной длиной последовательности используется функция `max()` с параметром `key=func_2`. После получения результата кэш очищается вызовом `func_2.cache_clear()`, чтобы не занимать память после завершения работы алгоритма.

### Алгоритм

1. Если текущее число равно `1`, функция возвращает длину `1`.
2. Для чётного числа выполняется переход к `n // 2`.
3. Для нечётного числа выполняется переход к `3n + 1`.
4. Результаты вычислений автоматически сохраняются в кэше.
5. Функция `max()` перебирает все числа от `1` до `n - 1` и выбирает то, для которого функция `func_2()` возвращает максимальную длину последовательности.
6. После завершения поиска кэш очищается вызовом `func_2.cache_clear()`.
7. Возвращается найденное начальное число.

### Отличия от предыдущего решения

- Используется декоратор `@cache` для сохранения уже вычисленных результатов.
- Повторяющиеся участки последовательностей вычисляются только один раз.
- Поиск максимума выполняется функцией `max()` с параметром `key`.
- После завершения работы кэш очищается методом `cache_clear()`.
- Производительность значительно возрастает за счёт устранения повторных вычислений.

### Вывод

Кэширование позволяет существенно ускорить поиск самой длинной последовательности Коллатца, поскольку одинаковые фрагменты последовательностей вычисляются только один раз. Очистка кэша после получения ответа освобождает память и делает функцию независимой от предыдущих запусков.

In [24]:
from functools import cache


@cache
def func_2(n):
    if n == 1:
        return 1
    if n % 2 == 0:
        return 1 + func_2(n // 2)
    return 1 + func_2(3 * n + 1)


def get_result_2(n: int) -> int:
    mx = max(range(1, n), key=func_2)
    func_2.cache_clear()
    return mx
    

start = perf_counter()
print(get_result_2(1_000_000))
end = perf_counter()

print(f'{end - start:.5f}')

837799
1.04804


## 3) Кэширование с объединением двух шагов

В данном решении сохраняется рекурсивный подход с кэшированием, однако обработка нечётных чисел дополнительно оптимизируется.

Если число `n` нечётное, по правилу Коллатца сначала выполняется переход:

$$
n \rightarrow 3n + 1
$$

Полученное число всегда является чётным, поэтому следующий шаг заранее известен:

$$
3n + 1 \rightarrow \frac{3n + 1}{2}
$$

Эти два перехода объединяются в один рекурсивный вызов:

```python
return 2 + func_3((3 * n + 1) // 2)

In [25]:
from functools import cache


@cache
def func_3(n):
    if n == 1:
        return 1
    if n % 2 == 0:
        return 1 + func_3(n // 2)
    return 2 + func_3((3 * n + 1) // 2)


def get_result_3(n: int) -> int:
    mx = max(range(1, n), key=func_3)
    func_3.cache_clear()
    return mx


start = perf_counter()
print(get_result_3(1_000_000))
end = perf_counter()

print(f'{end - start:.5f}')

837799
0.71577


## Сравнение времени выполнения

Для сравнения производительности используется функция `timeit()`, предназначенная специально для измерения времени выполнения небольших участков кода.

Каждый алгоритм запускается **10 раз**, после чего общее время делится на количество запусков. Таким образом получается **среднее время выполнения одного запуска**.

In [26]:
from timeit import timeit

funcs = (
    get_result_1,
    get_result_2,
    get_result_3,
)

for i, func in enumerate(funcs, start=1):
    tm = timeit(lambda: func(1_000_000), number=10) / 10
    print(f'Алгоритм {i}: {tm:.5f}')

Алгоритм 1: 12.68934
Алгоритм 2: 0.60513
Алгоритм 3: 0.45336


## Заключение

В ходе решения задачи были рассмотрены три различных подхода к поиску начального числа, порождающего самую длинную последовательность Коллатца.

Первый алгоритм использует рекурсивное построение последовательности без сохранения промежуточных результатов. Для каждого начального числа длина последовательности вычисляется заново, поэтому одинаковые участки последовательностей многократно пересчитываются. Из-за большого количества повторных вычислений этот подход оказался самым медленным.

Во втором решении применяется декоратор `@cache`, который автоматически сохраняет длины уже вычисленных последовательностей. Если алгоритм повторно достигает того же числа, результат берётся из кэша, что позволяет избежать повторных вычислений. После завершения работы кэш очищается вызовом `cache_clear()`.

Третий алгоритм дополнительно объединяет два последовательных шага для нечётных чисел. Поскольку после вычисления `3n + 1` всегда получается чётное число, следующий переход к `(3n + 1) / 2` выполняется сразу. Это уменьшает количество рекурсивных вызовов и дополнительно ускоряет работу алгоритма.

Среднее время выполнения алгоритмов:

- **Алгоритм 1:** `12.68934` с
- **Алгоритм 2:** `0.60513` с
- **Алгоритм 3:** `0.45336` с

Полученные результаты показывают, что основное ускорение достигается благодаря кэшированию. Использование `@cache` позволило сократить время выполнения более чем в **20 раз** по сравнению с базовой реализацией. Дополнительная оптимизация, объединяющая два шага для нечётных чисел, дала ещё около **25%** прироста производительности. В итоге третье решение оказалось самым эффективным среди рассмотренных.